### PR TITLE
test(share): add Full Node reconstruction tests

### PR DESCRIPTION
**Pull request by Wondertan**
Mainly:
* Improves testnet utility in share pkg
* Introduces tests that assert that our reconstruction works in semi-ideal scenarios with values taken from the original paper
	* 1x One Full and multiple Lights
	* 2x Two Fulls and two subnets with multiple Lights that cannot reconstruct data on their own
	
___For an in-depth description of each test consult a complementary comment to each test.___

Substitutes https://github.com/celestiaorg/celestia-node/pull/598
Based on #682 
Closes https://github.com/celestiaorg/celestia-node/issues/600

My favourite part about these tests is that they won't require any changes if we e.g., move to a completely new protocol for Share transportation

### CODE REVIEW

**Comment by adlerjohn** [comment]
Did you link to the right issue? #598 is a closed PR.

**Comment by adlerjohn** [comment]
Is this going to be merged into #682, or does it require #682 to be merged into `main` first? If the latter, please convert this into a draft to avoid merging too soon.

**Comment by Wondertan** [comment]
Draft until https://github.com/celestiaorg/celestia-node/pull/682 is merged

**Comment by Wondertan** [comment]
So the main blocker here is to make tests run while being limited with 8k goroutines, which is not trivial.

File: service/share/full_availability_test.go [hunk 1]
```diff
@@ -35,11 +38,54 @@ func TestShareAvailableOverMocknet_Full(t *testing.T) {
...
+	assert.NoError(t, err)
+}
+
+func TestShareAvailable_FullOverLights(t *testing.T) {
+	ipld.RetrieveQuadrantTimeout = time.Millisecond * 200
+	DefaultSampleAmount = 20 // s
+
+	const (
+		origSquareSize = 16 // k
+		lightNodes     = 69 // c
```
Comment (adlerjohn) [inline service/share/full_availability_test.go:56]: nice

File: service/share/testing.go [hunk 2]
```diff
@@ -74,63 +60,133 @@ func RandShares(t *testing.T, n int) []Share {
...
-	return NewService(bServ, NewLightAvailability(bServ)), root
+func (dn *dagNet) RandLightNode(n int) (*node, *Root) {
+	nd := dn.LightNode()
+	return nd, RandFillDAG(dn.t, n, nd.BlockService)
 }
 
-func (dn *DAGNet) RandFullService(n int) (*Service, *Root) {
-	bServ, root := dn.RandDAG(n)
-	return NewService(bServ, NewFullAvailability(bServ)), root
+func (dn *dagNet) RandFullNode(n int) (*node, *Root) {
```
Comment (adlerjohn) [inline service/share/testing.go:89]: While you're here, could you briefly document these? I wasn't actually sure what the parameter was until I read the implementation.

Comment (adlerjohn) [inline service/share/testing.go:89]: While you're here, could you briefly document these? I wasn't actually sure what the parameter was until I read the implementation.

File: service/share/full_availability_test.go (hunk 1 above)
Comment (Wondertan) [reply service/share/full_availability_test.go:53]: I updated this test and added more complex test scenarios. Check them out as well

**Comment by Wondertan** [comment]
Includes the code from https://github.com/celestiaorg/celestia-node/pull/815. Will be removed once it is merged

**Comment by Wondertan** [comment]
Unfortunately, it is very unlikely that we will see these tests merged with the race detector for them before upgrading to 1.19 with an unlimited race detector.

File: service/share/testing.go (hunk 2 above)
Comment (Wondertan) [reply service/share/testing.go:118]: Ok, sure

**Comment by Wondertan** [comment]
All the tests are passing reliably with go1.19beta1 with unlimited race detector. Run then 10 times in a row, now flakiness found yet

**Comment by liamsi** [comment]
What's the plan here? Disable race detector for this test till 1.19 is released? Or wait till 1.19 is released?

File: service/share/testing.go (hunk 2 above)
Comment (Wondertan) [reply service/share/testing.go:118]: I didn't comment this yet, so not sure why did you resolve @adlerjohn

Comment (Wondertan) [reply service/share/testing.go]: Updated now and commented on most of the code testing utilities that were not documented.

**Comment by Wondertan** [comment]
I extracted the tests and disabled the race detector for them. Fully ready for review

**Comment by Wondertan** [comment]
@musalbas, added you here as a reviewer. Do we need more harness tests or is this enough? Do we need tests where there are more than two Fulls with their own subnetworks?

**Comment by Wondertan** [comment]
Ok, then I think there is nothing else to do in this PR. Testground _reconstruction_ is planned but orthogonal to this PR.

**Review by renaynay — APPROVED** [review]

File: service/share/full_reconstruction_test.go [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+			}
+		}(i)
+	}
+
+	for i := 0; i < len(lights); i++ {
+		net.Connect(lights[i].ID(), source.ID())
+	}
+
+	for i := 0; i < len(lights); i++ {
+		net.Connect(lights[i].ID(), full.ID())
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply service/share/full_reconstruction_test.go:63]: Iirc, Mus wanted this to be in a specific order on the call. He wanted to find that if something is broken and we were also putting time.Sleeps with seconds between those for loops. Technically they could be unified. Don't think it matters

File: service/share/full_reconstruction_test.go [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+	"github.com/celestiaorg/celestia-node/ipld"
+)
+
+func init() {
+	ipld.RetrieveQuadrantTimeout = time.Millisecond * 100 // to speed up tests
+}
+
+// TestShareAvailable_OneFullNode asserts that a FullAvailability node can ensure
+// data is available(reconstruct data square) while being connected to
+// LightAvailability node's only.
```
Comment (adlerjohn) [inline service/share/full_reconstruction_test.go:22]:
```suggestion
// TestShareAvailable_OneFullNode asserts that a FullAvailability node can ensure
// data is available (reconstruct data square) while being connected to
// LightAvailability nodes only.
```

**Review by adlerjohn — APPROVED** [review]

**Pull request merged**